### ISSUE TITLE
Extending the model should follow the DRY principle

### ISSUE THREAD

**Issue by PeterWippermann** [niloc132/gxt-driver#11]
Today I noticed that `org.senchalabs.gwt.gwtdriver.gxt.models.Field` only lets you send keys, but doesn't support to clear a field. Besides the fact that it would be certainly nice to implement this in the original model (would you like to have a pull request for this?), I decided to write my own model and implement it there.

Writing a model isn't very hard, but I think there's a lot of unnecessary repitition - which violates the DRY principle. So I propose you make the model classes more easily extendable.

For example, please see my current implementation below:

``` java

package com.acme.app.gui.tests.model.driver;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.senchalabs.gwt.gwtdriver.gxt.models.Field.FieldFinder;
import org.senchalabs.gwt.gwtdriver.models.GwtWidget;
import org.senchalabs.gwt.gwtdriver.models.GwtWidgetFinder;

import com.acme.app.gui.tests.model.driver.ExtendedGxtField.ExtendedGxtFieldFinder;

public class ExtendedGxtField extends GwtWidget<ExtendedGxtFieldFinder> {

    public ExtendedGxtField(WebDriver driver, WebElement element) {
        super(driver, element);

    }

    public WebElement getInputElement() {
        return getElement().findElement(By.tagName("input"));
    }

    public void clear() {
        getInputElement().clear();
    }

    public static class ExtendedGxtFieldFinder extends GwtWidgetFinder<ExtendedGxtField> {
        private FieldFinder fieldFinder = new FieldFinder();

        public ExtendedGxtFieldFinder withText(String text) {
            fieldFinder.withText(text);
            return this;
        }

        public ExtendedGxtFieldFinder withLabel(String label) {
            fieldFinder.withLabel(label);
            return this;
        }

        @Override
        public ExtendedGxtFieldFinder withDriver(WebDriver driver) {
            fieldFinder.withDriver(driver);
            return this;
        }

        @Override
        public ExtendedGxtField done() {
            return new ExtendedGxtField(driver, fieldFinder.done().getElement());
        }
    }
}
```
- I wanted to extend `org.senchalabs.gwt.gwtdriver.gxt.models.Field`, which would have been very nice I think. But this isn't possible, because the parameterized methods yield to a conflict between mine and the super class. 
- Maybe `GwtWidget<?>` should be a separate interface? You can still provide an abstract implementation, which the concrete models can extend.
- Why is the logic to find the encapsulated, nested elements hidden in den model classes? I propose all models expose the nested `WebElement` like I did with `getInputElement()`.
- I also would have liked to reuse `FieldFinder`, but because of the generics this wasn't possible also. Instead I had to implement my own finder and delegate all calls.

**Comment by niloc132** [comment]
Its a nice idea, but covariance/contravariance makes most of these impossible or cumbersome.

You didn't need to make a new finder I think, but just extend the existing one, and reimplement `done()` with a tighter return type. If desired, override each method and make it call super, then cast the returned value to your own finder - this seems more appropriate than your facade use, and I'm not certain why you didn't take this approach.

The alternative would be for every finder `F` to be generic on the model `M`, so that you have `F<M>`, and each method returns `F<M>`. In turn, since any widget is generic on its Finder (granting easy access to a finder instance when GwtWidget.find(M.class) is called), we already have `M` generic on `F`. Now with your suggestion, we have `GwtWidget<M extends GwtWidget<M, F>, F extends GwtWidgetFinder<M, F>`. What does this mean? First step, just use this in Field and FieldFinder:

```
public class Field extends GwtWidget<Field, FieldFinder> {
  //...
  public static class FieldFinder extends GwtWidgetFinder<Field, FieldFinder> {
    //...
    public FieldFinder withLabel(String label) {
      //...
    }
    public Field done() {
      //...
    }
  }
}
```

Simple usage (obviously not complete, but lets just poke the API quick):

```
Field f = GwtWidget.find(Field.class).withLabel("Name").done();
```

Okay, so lets replace any FieldFinder and Field instance in there with the generics, so that Field and FieldFinder can be appropriately subclassed as you describe:

```
public class Field<M extends Field<M, F>, F extends FieldFinder<M, F>> extends GwtWidget<M, F> {
  //...
  public static class FieldFinder<M extends Field<M, F>, F extends FieldFinder<M, F>> extends GwtWidgetFinder<M, F> {
    //...
    public F withLabel(String label) {
      //...
    }
    public M done() {
      //...
    }
  }
}
```

Okay, so that's a bit messier, but at least we don't need to override `withLabel` any more. Obviously we still need to override `done()` to change the returned value to the correct type (and I'm glossing over extra correctness of use of `super` and `extends`, see the generics chapter of Effective Java 2nd edition). Now how do we use it?

```
Field<Field, FieldFinder> f = GwtWidget.find(Field.class).withLabel("Name").done();
```

Okay, so thats a bit hideous, and is stuck on _every widget model you ever declare_, and every variable that is assignable to `GwtWidget<M, F>`, going forward, forever - technically we can skip those generics and be raw everywhere we actually use a field, but we're throwing away type safety in a few cases in the name of DRY, and we still get to repeat plenty.

Of course, if we just got rid of that pesky GwtWidget.find method, then we don't have any of this recursive generics crap, at least not on GwtWidget. Still need it on the finder type, or your builder methods wont work correctly...

---

I welcome a concrete suggestion that manages this without taking away features that already exist. Keep the GwtWidget.as(Class) method in mind, which lets you say "I found this widget already, but lets use that model over there to interact with it". 

Also, pull requests are welcome to add methods, but be _very_ careful with `WebElement.clear()`, which doesn't actually cause a DOM event, so the widget won't notice - this is the equivelent of being on a webpage and rather than clicking on anything, you modify the dom directly, and just keep hoping that all use cases work perfectly. I don't expect it to consistently work with GXT or any library that is monitoring events, especially for complex widgets like the ComboBox or DateField.

**Comment by niloc132** [comment]
> Why is the logic to find the encapsulated, nested elements hidden in den model classes? I propose all models expose the nested WebElement like I did with getInputElement().

This might be reasonable, but would be very easy to overdo and provide a bunch of APIs that don't make any sense. For example, finding an `<input>` will fail on a `Textarea`, and any API that encourages directly manipulating the DOM is going to have issues with alternate Appearance implementations. 

In any API design, one strives to expose what must be used, and of course open up more as necessary, but if you take this to an extreme, all fields and methods are `public`, and `private` or `protected` should never have existed in Java in the first place.

**Comment by PeterWippermann** [comment]
@niloc132 regarding my proposal for a "getInputElement()" method. I thought one could do this for every element, where this makes sense. Of course a Textarea should return an HTML-textarea and not an HTML-input.

However, looking at the GXTdriver model classes, this only seems to make sense for the Field class at all, since it's the only one wrapping a (nested) native HTML element. One could also imagine this for the Button, but I see no usecase there...

**Comment by PeterWippermann** [comment]
Extending the existing `Field` and `FinderFinder` class doesn't work also. Even I use a tighter return type, that one get's hidden: The return type (visible from for caller) of done() is `Field`, even if I return my `ExtendedGxtField`.

Don't you think a separate interface could work?
`ExtendedGxtField extends Field implements GwtWidget<GwtWidgetFinder<ExtendedGxtField>>`
Or do we get a problem with interface ambiguity then? (Because Field also implements the interface with a different parameter type)

**Comment by niloc132** [comment]
> Or do we get a problem with interface ambiguity then? (Because Field also implements the interface with a different parameter type)

Right, thats how GwtWidget.find(Class) works - it uses the finder declared in the class you pass in. If you pass in `Field` which `extends GwtWidget<FieldFinder>`, then it will return a FieldFinder for you to work with. Try it, what you have there isn't legal Java, since it implements two differently-generic version of the same interface (do it with `List<String>` and `List<Double>` for example).

Look into `GwtWidget.as`, it sounds to be exactly what you are looking for (but not asking for) to find a Field and then use it as some other class (possibly not even a subclass) instead.

With that said, if you do come up with a model/finder idea that doesn't have that mess of generics, doesn't require any `as(Class)` calls, and isn't more complex than we currently have, I'd be willing to consider it even if it is a breaking change, though obviously we'd need to do a major release to redo it. I'll help with rebuilding into something new, but it really has to meet all of these criteria - making it slightly easier to add util methods but making all of your code uglier and harder to write would entirely defeat the purpose of the tool.

**Comment by niloc132** [comment]
> However, looking at the GXTdriver model classes, this only seems to make sense for the Field class at all, since it's the only one wrapping a (nested) native HTML element.

Not sure I follow this statement - some fields are rather complex, such as ComboBox and DateField. However, given the complete control that an appearance implementation has over their dom structure, it is also essentially impossible to make assumptions about how the widget's dom is put together, hence the clientmethods interface/wiring.
